### PR TITLE
Welsh preprocessing + one change to common.py

### PR DESCRIPTION
**Pull request by JRMeyer**
1) added a whitespace collapsing line in common.py, which will replace all whitespace(s) with a single space. This should also capture tabs and non-standard whitespaces.
2) added to welsh pre-processor, to check for normal letters

### CODE REVIEW

File: src/corporacreator/preprocessors/cy.py [hunk 1]
```diff
@@ -9,4 +9,23 @@ def cy(client_id, sentence):
...
+                                         'm', 'n', 'o', 'ó', 'ò', 'ô', 'ö', 'p',
+                                         'r', 's', 't', 'u', 'ú', 'ù', 'û', 'ü',
+                                         'w', 'ẃ', 'ẁ', 'ŵ', 'ẅ', 'y', 'ý', 'ỳ',
+                                         'ŷ', 'ÿ'])):
+        pass
+    else:
+        # Not all chars are in the Welsh alphabet!
+        # more processing needed here!
+        pass
+
```
Comment (kdavis-mozilla) [inline src/corporacreator/preprocessors/cy.py]: There are several issues:

- Could you do something in the "else" case?
- Also, `sentence` is cased, but your alphabet is not.
- Also you re-create the set `set(['' ', 'a','á', 'à', 'â',...])` for each sentence. You should create is once and reuse it
- You don't need to do `set(list(sentence))` you only need `set(sentence)`
- This should be moved in to a function and not be inline in the `cy` function

All of this is also true of all the other PR's too.

File: src/corporacreator/preprocessors/common.py [hunk 2]
```diff
@@ -75,5 +75,7 @@ def common(sentence):
     sentence = _strip_tags(sentence)
     # Remove non-printable characters
     sentence = _strip_string(sentence)
+    # collapse all whitespace and replace with single space
+    sentence = (' ').join(sentence.split())
```
[the remark this answers is not in the document]
Comment (JRMeyer) [reply src/corporacreator/preprocessors/common.py]: addressed in #80

File: src/corporacreator/preprocessors/cy.py (hunk 1 above)
Comment (JRMeyer) [reply src/corporacreator/preprocessors/cy.py]: @kdavis-mozilla all your concerns have been addressed in newest commits to this branch. 

The only thing that I did different was using not a function to compile the `set()` of the alphabet, but just a single, first line in the file. I'm not sure of a better way to do this with a function.

File: src/corporacreator/preprocessors/cy.py [hunk 3]
```diff
@@ -1,12 +1,38 @@
-def cy(client_id, sentence):
-    """Cleans up the passed sentence, removing or reformatting invalid data.
+# compile the set of the alphabet just once
+alphabet = set([' ', 'a','á', 'à', 'â', 'ä', 'b', 'c',
+                'd', 'e', 'é', 'è', 'ê', 'ë', 'f', 'g',
+                'h', 'i', 'í', 'ì', 'î', 'ï', 'j', 'l',
+                'm', 'n', 'o', 'ó', 'ò', 'ô', 'ö', 'p',
+                'r', 's', 't', 'u', 'ú', 'ù', 'û', 'ü',
+                'w', 'ẃ', 'ẁ', 'ŵ', 'ẅ', 'y', 'ý', 'ỳ',
+                'ŷ', 'ÿ'])
```
Comment (kdavis-mozilla) [inline src/corporacreator/preprocessors/cy.py]: Should include upper and lower case letters.

File: src/corporacreator/preprocessors/cy.py [hunk 4]
```diff
@@ -1,12 +1,38 @@
...
 
+def cy(client_id, sentence):
+    """
+    Cleans up the passed sentence, removing or reformatting invalid data.
     Args:
       client_id (str): Client ID of sentence's speaker
       sentence (str): Sentence to be cleaned up.
+      alphabet (set): set of unique, lowercased characters in language.
```
Comment (kdavis-mozilla) [inline src/corporacreator/preprocessors/cy.py]: This should only document the arguments to the function `cy()`, not the global variable `alphabet`.

File: src/corporacreator/preprocessors/cy.py [hunk 5]
```diff
@@ -1,12 +1,38 @@
...
+    # the sentence is purely Welsh characters as per https://github.com/
+    # unicode-cldr/cldr-misc-full/blob/master/main/cy/characters.json
+    if set(sentence.lower()).issubset(alphabet):
+        pass
+    else:
+        # Not all chars are in the Welsh alphabet!
+        # more processing needed here!
+        sentence = sentence.replace(",", "") # remove commas
+        sentence = sentence.replace("’", "'") # fix apostrophes
+        sentence = sentence.replace("wwna", "wna")
```
Comment (kdavis-mozilla) [inline src/corporacreator/preprocessors/cy.py]: All characters of "wwna" are in the alphabet, but this is only run on sentence if not all characters in the sentence are in alphabet.

So this is only run on sentences that have character not in alphabet **_and_** also contain the string "wwna".  So in particular, it does not have same effect as Dewi's code.

File: src/corporacreator/preprocessors/cy.py [hunk 6]
```diff
@@ -1,12 +1,38 @@
...
+
+    # this if loop will skip the following else loop in the case that
+    # the sentence is purely Welsh characters as per https://github.com/
+    # unicode-cldr/cldr-misc-full/blob/master/main/cy/characters.json
+    if set(sentence.lower()).issubset(alphabet):
+        pass
+    else:
+        # Not all chars are in the Welsh alphabet!
+        # more processing needed here!
+        sentence = sentence.replace(",", "") # remove commas
```
Comment (kdavis-mozilla) [inline src/corporacreator/preprocessors/cy.py]: Why remove commas? The data set will be used in more than Deep Speech. So we should leave in as much information as possible.

File: src/corporacreator/preprocessors/cy.py [hunk 7]
```diff
@@ -1,12 +1,38 @@
...
+    # alphabet is compiled at the beginning of this script
+    if set(sentence.lower()).issubset(alphabet):
+        pass
+    else:
+        # Not all chars are in the Welsh alphabet!
+        # more processing needed here!
+        sentence = sentence.replace(",", "") # remove commas
+        sentence = sentence.replace("’", "'") # fix apostrophes
+        sentence = sentence.replace("wwna", "wna")
+        sentence = sentence.replace(" siwr ", " siŵr ")
```
Comment (kdavis-mozilla) [inline src/corporacreator/preprocessors/cy.py]: Ditto what I mentioned above here

File: src/corporacreator/preprocessors/cy.py (hunk 3 above)
Comment (JRMeyer) [reply src/corporacreator/preprocessors/cy.py]: I would vote that we keep the alphabet lowercase, and then lowercase the sentence when 
we check whether or not the sentence is a subset of the alphabet, as in here:

```
     if set(sentence.lower()).issubset(alphabet):
```

To me, this seems easier for a contributer to be able to quickly read the alphabet.

File: src/corporacreator/preprocessors/cy.py (hunk 6 above)
Comment (JRMeyer) [reply src/corporacreator/preprocessors/cy.py]: @kdavis-mozilla I think we are thinking about the use-cases of CorporaCreator differently. I see this as a tool that can be easily adjusted for each engineer's use case. So, if someone wants to take out the comma, then that's great, but otherwise they can leave it in. @DewiBrynJones might need commas for text applications, but for speech (like English) it might make more sense to remove them.

At this stage, this isn't a definitive text processor for Welsh, and it will keep growing. 

In short, I'm not too concerned about commas staying in or not, because the engineer who
will be using this (@DewiBrynJones and others) can get their own `clips.tsv` and run their
own processing. 

Thoughts?

File: src/corporacreator/preprocessors/cy.py (hunk 5 above)
Comment (JRMeyer) [reply src/corporacreator/preprocessors/cy.py]: @kdavis-mozilla this is an issue, indeed... suggestions?

File: src/corporacreator/preprocessors/cy.py (hunk 6 above)
Comment (DewiBrynJones) [reply src/corporacreator/preprocessors/cy.py]: I can handle removing commas in my own code. (and other symbols) 

https://github.com/DewiBrynJones/docker-deepspeech-cy/blob/commonvoice_data/local/bin/language_modelling_utils.py#L37

it's good to know that you may be considering CorpusCreator for other (non-speech) use cases.  I'd vote for the flexibilty despite being just speech for us for the moment. 

(_we were considering how to handle English words that are not pronounced to a strictly Welsh phone set. (e.g. 'celcius' is pronunced/replaced with 'selsiws', 'wallace' as 'walus'. This could be handled then local speech specific scripts_ )

File: src/corporacreator/preprocessors/cy.py (hunk 3 above)
Comment (kdavis-mozilla) [reply src/corporacreator/preprocessors/cy.py]: Casing is local specific, e.g. [Turkish](https://en.wikipedia.org/wiki/Dotted_and_dotless_I).

So doing what you suggest would mean that we have to specify which local the code should be run in. We should just have to run the code and have it work.

File: src/corporacreator/preprocessors/cy.py (hunk 6 above)
Comment (kdavis-mozilla) [reply src/corporacreator/preprocessors/cy.py]: CorporaCreator is to to create the release data set, not a particular data set for a particular engineer.

So we should leave in any information that's correct and remove any information that is not. Doing so will allow the release data set to be of the most use to the greatest number of people.

We aren't just writing this for Dewi's use case or our use case. It's for everyone's use cases.

File: src/corporacreator/preprocessors/cy.py (hunk 5 above)
Comment (kdavis-mozilla) [reply src/corporacreator/preprocessors/cy.py]: Move this to the other branch of the else and remove the pass.

File: src/corporacreator/preprocessors/cy.py (hunk 6 above)
Comment (DewiBrynJones) [reply src/corporacreator/preprocessors/cy.py]: > We aren't just writing this for Dewi's use case or our use case. It's for everyone's use cases.

Of course! 

My use case aligns with this - to make it as easy as possible for anyone to use the Welsh data to train DeepSpeech. 

If you envision the corpora being used for non-speech use cases (by anyone), then some further 'corrections' I was thinking of won't be relevant. (see earlier comment) .

File: src/corporacreator/preprocessors/cy.py (hunk 3 above)
Comment (JRMeyer) [reply src/corporacreator/preprocessors/cy.py]: @kdavis-mozilla --- lowercasing removed in latest commit

File: src/corporacreator/preprocessors/cy.py (hunk 5 above)
Comment (JRMeyer) [reply src/corporacreator/preprocessors/cy.py]: resolved in latest commit, no alphabet checking loop

File: src/corporacreator/preprocessors/cy.py (hunk 4 above)
Comment (JRMeyer) [reply src/corporacreator/preprocessors/cy.py]: resolved in latest commit

File: src/corporacreator/preprocessors/cy.py (hunk 6 above)
Comment (JRMeyer) [reply src/corporacreator/preprocessors/cy.py]: resolved in latest commit, took out commas. From now on, I will only remove strictly incorrect characters / strings.

File: src/corporacreator/preprocessors/cy.py (hunk 1 above)
Comment (JRMeyer) [reply src/corporacreator/preprocessors/cy.py]: resolved in latest commit. No more alphabet-checking.

File: src/corporacreator/preprocessors/cy.py (hunk 7 above)
Comment (JRMeyer) [reply src/corporacreator/preprocessors/cy.py]: resolved in latest commit.

no looping at all, so these `.replace()` methods are used on every sentence.
